nixos/modules/services/security/esdm.nix

```diff
@@ -0,0 +1,137 @@
+    enable = lib.mkEnableOption (lib.mdDoc "ESDM service configuration");
+    package = lib.mkPackageOptionMD pkgs "esdm" { };
+    kernelEnable = lib.mkOption {
+      type = lib.types.bool;
+      default = false;
+      description = lib.mdDoc ''
+        enable ESDM kernel module
+      '';
+    };
+    serverEnable = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
+      description = lib.mdDoc ''
+        enable option for ESDM server service
+      '';
+    };
+    cuseRandomEnable = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
+      description = lib.mdDoc ''
+        enable option for ESDM cuse-random service
+      '';
+    };
+    cuseUrandomEnable = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
+      description = lib.mdDoc ''
+        enable option for ESDM cuse-urandom service
+      '';
+    };
+    procEnable = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
+      description = lib.mdDoc ''
+        enable option for ESDM proc service
+      '';
+    };
+    verbose = lib.mkOption {
+      type = lib.types.bool;
+      default = false;
+      description = lib.mdDoc ''
+        enable verbose ExecStart for ESDM
+      '';
+    };
+    esdmHashName = lib.mkOption {
+      type = lib.types.str;
+      default = "sha3-512";
+      description = lib.mdDoc ''
+        hash configration of the kernel module esdm_es
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable (
+    lib.mkMerge [
+      ({
+        systemd.packages = [ cfg.package ];
+
+        assertions = lib.lists.singleton {
+          assertion = (linux_6_3 || cfg.kernelEnable) == true;
+          message = "Expected kernel-version >= 6.3. got kernel-version:${kernelVersion}";
+        };
+      })
+      (lib.mkIf cfg.kernelEnable {
+        boot.extraModulePackages = [ config.boot.kernelPackages.esdm_es.out ];
+        boot.extraModprobeConfig = ''
+          options esdm_es esdm_hash_name=${cfg.esdmHashName}
+        '';
+        boot.kernelModules = [ "esdm_es" ];
+
+        #patch kernel (works for kernel version 6.3)
+        boot.kernelPatches = [ ]
+          ++ lib.lists.optionals linux_6_3 [
+          {
+            name = "esdm_sched_es_hook";
+            patch = "${cfg.package}/addon/linux_esdm_es/0001-ESDM-scheduler-entropy-source-hooks_6.4.patch";
+          }
+          {
+            name = "esdm_inter_es_hook";
+            patch = "${cfg.package}/addon/linux_esdm_es/0002-ESDM-interrupt-entropy-source-hooks_6.4.patch";
+          }
+        ];
+      })
+
+      #need the following lines for NixOS to recongnize, that it indeed should start these services
+      (lib.mkIf cfg.serverEnable {
+        systemd.services."esdm-server".wantedBy = [ "basic.target" ];
+        systemd.services."esdm-server".serviceConfig = lib.mkIf cfg.verbose {
+          ExecStart = [
+            " " # unset previous value defined in 'esdm-server.service.in'
```

Comment:
```suggestion
            " " # unset previous value defined in 'esdm-server.service.service'
```

Reply:
This should be 'esdm-server.service' because that's the name of the file in the output of the ESDM package (and also the canonical name for a systemd service).

Reply:
It is now changed to 'esdm-server.service' and so are the other 3 lines.